pkg/sparrow/targets/remote/gitlab/gitlab.go

```diff
@@ -56,184 +53,202 @@ type Config struct {
-		log.Error("Failed to fetch file list", "status", resp.Status)
+		log.ErrorContext(ctx, "Failed to fetch file list", "status", resp.Status)
 		return nil, fmt.Errorf("request failed, status is %s", resp.Status)
 	}
 
 	var fl []file
 	err = json.NewDecoder(resp.Body).Decode(&fl)
 	if err != nil {
-		log.Error("Failed to decode file list", "error", err)
+		log.ErrorContext(ctx, "Failed to decode file list", "error", err)
 		return nil, err
 	}
 
-	var result []string
 	for _, f := range fl {
 		if strings.HasSuffix(f.Name, ".json") {
-			result = append(result, f.Name)
+			files = append(files, f.Name)
 		}
 	}
 
-	log.Debug("Successfully fetched file list", "files", len(result))
-	return result, nil
+	log.DebugContext(ctx, "Successfully fetched file list", "files", len(files))
+	return files, nil
 }
 
 // PutFile commits the current instance to the configured gitlab repository
 // as a global target for other sparrow instances to discover
-func (c *client) PutFile(ctx context.Context, body remote.File) error { //nolint: dupl,gocritic // no need to refactor yet
+func (c *client) PutFile(ctx context.Context, file remote.File) (err error) { //nolint: dupl,gocritic // no need to refactor yet
 	log := logger.FromContext(ctx)
-	log.Debug("Registering sparrow instance to gitlab")
+	log.DebugContext(ctx, "Registering sparrow instance to gitlab")
 
 	// chose method based on whether the registration has already happened
-	n := url.PathEscape(body.Name)
-	b, err := body.Bytes()
+	n := url.PathEscape(file.Name)
+	b, err := file.Serialize(c.config.Branch)
 	if err != nil {
-		log.Error("Failed to create request", "error", err)
+		log.ErrorContext(ctx, "Failed to create request", "error", err)
 		return err
 	}
 	req, err := http.NewRequestWithContext(ctx,
 		http.MethodPut,
-		fmt.Sprintf("%s/api/v4/projects/%d/repository/files/%s", c.baseUrl, c.projectID, n),
+		fmt.Sprintf("%s/api/v4/projects/%d/repository/files/%s", c.config.BaseURL, c.config.ProjectID, n),
 		bytes.NewBuffer(b),
 	)
 	if err != nil {
-		log.Error("Failed to create request", "error", err)
+		log.ErrorContext(ctx, "Failed to create request", "error", err)
 		return err
 	}
 
-	req.Header.Add("PRIVATE-TOKEN", c.token)
+	req.Header.Add("PRIVATE-TOKEN", c.config.Token)
 	req.Header.Add("Content-Type", "application/json")
 
-	resp, err := c.client.Do(req) //nolint:bodyclose // closed in defer
+	resp, err := c.client.Do(req)
 	if err != nil {
-		log.Error("Failed to push registration file", "error", err)
+		log.ErrorContext(ctx, "Failed to push registration file", "error", err)
 		return err
 	}
 
-	defer func(Body io.ReadCloser) {
-		err = Body.Close()
-		if err != nil {
-			log.Error("Failed to close response body", "error", err)
+	defer func() {
+		cErr := resp.Body.Close()
+		if cErr != nil {
+			log.ErrorContext(ctx, "Failed to close response body", "error", cErr)
+			err = errors.Join(err, cErr)
 		}
-	}(resp.Body)
+	}()
+
+	if resp.StatusCode == http.StatusBadRequest {
```

Comment:
I've implemented this because @niklastreml and I debugged this about 2h yesterday and the reason was exactly this case I'm trying to cover here. It is also the reason why the issue exists in the first place.

Unfortunately the Gitlab API does not give any information about what went wrong and just returns a `400 Bad Request` which is very vague.

Imo we should keep this log because it gives the user a starting point to debug and not just leaves him with the rather generic error message.

If you feel strongly about removing this I'd propose add a link to the log message to the [Gitlab API docs](https://docs.gitlab.com/ee/api/repository_files.html#update-existing-file-in-repository) so the user can see some possible reasons this error happened. But again unfortunately the Gitlab API docs don't cover every case this error can happen so I'd still keep the part about the protected branch in this log.

> If the commit fails for any reason we return a 400 Bad Request error with a non-specific error message. Possible causes for a failed commit include:
> 
> The file_path contained /../ (attempted directory traversal).
> The commit was empty: new file contents were identical to the current file contents.
> Someone updated the branch with git push while the file edit was in progress.
>
> ~ _GitLab API documentation - Update existing file in repository_

Reply:
I'm sorry that you had to debug for two hours but like mentioned, this is a temporal implementation detail of how gitlab works. The developer should know that they should refer to the gitlab documentation regarding error handling.

It's neither our responsibility to handle every single possible case (why aren't we then adding additional cases for other error codes? Just because they haven't happened yet doesn't mean they won't happen). Just logging an assumption (that is currently true for one use case) may be helpful, but where do you even draw the line? It's impossible.

Just log the error and the error code. It's not our job to interpret every single special response of the gitlab API because someone decided to block the main branch of every repository. 😉 This makes the code non consistent.

I'm not opposed to adding a comment in the general error catching block linking to the gitlab API docs.

Reply:
I've added comments instead of special error case handling to give a debugging developer a clue about what could have possibly went wrong.